test-support/reference-trie/src/lib.rs

```diff
@@ -585,6 +557,57 @@ fn partial_encode(partial: Partial, node_kind: NodeKindNoExt) -> Vec<u8> {
 	output
 }
 
+struct ByteSliceInput<'a> {
```

Comment:
- take method now uses a struct (butesliceinput)
  -> do we want to move this bytesliceinput in a common crate: take was already probably copied between implementations, but this is a bit more code.
  I would suggest to put it nearby the Codec, and NodePlan place.
- Or could we use std 'Cursor' instead of bytesliceinput.
- decode plan could use 'cursor', 'byteslice' or 'read' trait implementation as a parameter.
 Read trait is my favorite as it can extend to nocopy proof decoding.


- Or could we use std 'Cursor' instead of bytesliceinput.

Reply:
Definitely there's too much duplication between `NodeCodec` implementations, but I'd rather leave the task of dealing with that and refactoring into utilities for another PR.

Reply:
Do you think 'Cursor' could be use instead? Iirc you could not yes mut ref over slice because it could not go backward, but set_position or seek should be fine.

Reply:
I think it could work, but `Cursor` is in `std::io` which is not imported into the Substrate runtime with `no_std`. And we will want the trie codecs to work inside the runtime. Also, `ByteSliceInput` is a pretty small amount of code IMO.

Reply:
Oh yes, I ovelooked that, Read, Write and Cursor should move to alloc crate in my opinion (afaik only box is responsible for it being in io).